Add threads pool tag scanning

RuBublik opened the pull request:
## Pull Request: Adding support for pull-tag scanning of _EThread structures and windows thrdscan plugin

This pull request aims to enhance the functionality of Volatility by adding support for pull-tag scanning of _EThread structures in Windows memory. Additionally, it introduces a new plugin called `windows.thrdscan.ThrdScan` that utilizes the newly added features.

The changes included in this pull request are as follows:

- Addition of constraint definitions for thread objects.
- Extension of the `ETHREAD` class in the `pool` module to support pull-tag scanning.
- Implementation of `thrdscan`, which is the equivalent of the same plugin for Volatility 2.
- Implementation of a test for the new plugin in test_volatility.py module.

This addition enables efficient extraction of thread-related information from memory images, addressing a gap in Volatility 3. It also introduces another method for memory forensic analysts to potentially discover hidden or malicious processes by leveraging the relationship between _EThread and its parent _EProcess, similar to the functionality provided by the `psxview` plugin in Volatility 2.

## Example

Here is a sample output from running the `windows.thrdscan.ThrdScan` plugin on a Stuxnet memory image:

```
C:\volatility3>python vol.py -f "C:\mem_images\stuxnet.vmem" windows.thrdscan.ThrdScan    
Volatility 3 Framework 2.4.2
Progress:  100.00               PDB scanning finished
Offset  PID     TID     Start Address   Create Time     Exit Time

0x5c206510      2460    2468    0x77667098      2017-08-04 19:51:41
0x5c208030      2460    2480    0x77667098      2017-08-04 19:51:41
0x5c20a7c8      1072    3208    0x77667098      2012-06-01 19:57:40
0x5c214d48      1176    3308    0x77667098      2017-08-04 20:00:01
0x5c2159a8      880     2500    0x77667098      2017-08-04 19:51:41     2012-06-04 20:01:25
0x5c216d48      2460    2496    0x77667098      2017-08-04 19:51:41
0x5c222238      2460    2512    0x883f21a0      2017-08-04 19:51:41
0x5c222d48      2460    2504    0x77667098      2017-08-04 19:51:41
0x5c224030      2460    2516    0x77667098      2017-08-04 19:51:41
0x5c224588      2460    2524    0x77667098      2017-08-04 19:51:41
0x5c224a28      2460    2520    0x77667098      2017-08-04 19:51:41
0x5c232398      2064    2536    0x930e3180      2017-08-04 19:51:41
0x5c235cc8      2460    1940    0x77667098      2017-08-04 19:59:42
0x5c238030      1176    2936    0x77667098      2017-08-04 19:53:19
0x5c24b648      1772    2624    0x77667098      2017-08-04 19:51:41
0x5c253638      2064    2636    0x77667098      2017-08-04 19:51:42
0x5c2558e0      4       2724    0x82a669c2      2017-08-04 19:52:27
0x5c255bc8      2064    2640    0x77667098      2017-08-04 19:51:42
0x5c258348      2012    2644    0x77667098      2012-06-04 20:01:13
0x5c25a030      2236    2692    0x77667098      2017-08-04 19:52:05
0x5c402370      1072    1980    0x77667098      2017-08-04 19:51:28
0x5c4073a0      292     984     0x77667098      2017-08-04 19:51:28
0x5c40a030      912     1512    0x77667098      2017-08-04 19:51:13
0x5c40c618      1308    1876    0x77667098      2012-06-04 20:01:20
0x5c40d430      912     1296    0x77667098      2017-08-04 19:51:34
0x5c40e348      912     4032    0x77667098      2017-08-04 19:56:21     2012-06-01 19:57:35
0x5c4104c0      880     1288    0x77667098      2017-08-04 19:51:34     2017-08-04 19:51:34
0x5c41d150      1176    1520    0x77667098      2017-08-04 19:51:13
0x5c422350      912     1780    0x77667098      2012-06-01 19:57:01
0x5c423b20      1176    1524    0x77667098      2017-08-04 19:51:13     2012-06-04 20:01:25
0x5c424418      1176    1536    0x8c9425b8      2017-08-04 19:51:13
0x5c426bf0      532     3216    0x77667098      2012-06-01 19:57:42
0x5c42ba20      2012    3596    0x77667098      2017-08-04 20:00:01
0x5c443a68      712     2308    0x77667098      2017-08-04 19:56:29
0x5c449398      1176    3312    0x77667098      2017-08-04 20:00:01
0x5c449680      2764    2968    0x77667098      2017-08-04 19:53:19
0x5c449d48      1176    3792    0x96c37b00      2017-08-04 19:56:17
0x5c44a030      912     1572    0x77667098      2017-08-04 19:51:13
0x5c44a778      516     2204    0x77667098      2012-06-04 20:00:50
0x5c44aa60      912     1592    0x91c2c7d0      2017-08-04 19:51:13
0x5c44ad48      912     3224    0x77667098      2017-08-04 19:59:50
0x5c44ed48      1072    3052    0x77667098      2017-08-04 20:00:02
0x5c4588c0      528     308     0x81e34a48      2017-08-04 19:50:22
0x5c459030      912     1600    0x77667098      2017-08-04 19:51:13     2017-08-04 19:53:19
0x5c459490      912     1616    0x77667098      2017-08-04 19:51:13     2012-06-04 20:01:25
0x5c459a60      912     1608    0x77667098      2017-08-04 19:51:13     2017-08-04 19:53:19
0x5c45aa30      1392    2040    0x77667098      2012-06-04 20:00:57
0x5c45b030      912     1640    0x77667098      2017-08-04 19:51:13
0x5c45c7d0      912     1628    0x85a5c9e4      2017-08-04 19:51:13     2017-08-04 19:54:36
0x5c462a70      912     1516    0x8c9376c0      2017-08-04 19:59:29
0x5c462d48      4       1644    0x94911115      2017-08-04 19:51:13
0x5c463020      4       1656    0x94911115      2017-08-04 19:51:13
0x5c464030      3064    3352    0x77667098      2017-08-04 19:58:34
0x5c464d48      2244    2688    0x77667098      2017-08-04 19:52:05
0x5c465518      1392    1760    0x77667098      2017-08-04 19:51:13
0x5c467c48      912     1428    0x77667098      2017-08-04 19:59:19
0x5c468c50      4       1716    0x9496f7c6      2017-08-04 19:51:13
0x5c469340      4       1692    0x94911115      2017-08-04 19:51:13
0x5c469d48      2064    2428    0x77667098      2017-08-04 19:51:40
0x5c46b848      4       1708    0x9496f7c6      2017-08-04 19:51:13
0x5c46c030      308     2856    0x77667098      2012-06-04 20:00:37
0x5c46c8a8      912     1684    0x77667098      2017-08-04 19:51:13
0x5c46cd48      4       1672    0x94911115      2017-08-04 19:51:13
0x5c46d768      4       1704    0x9496f7c6      2017-08-04 19:51:13
0x5c46e8b0      4       1692    0x92f9f7c6      2017-08-04 19:48:16
0x5c46fd48      4       1700    0x94911115      2017-08-04 19:51:13
0x5c475160      912     2336    0x77667098      2012-06-01 19:57:57
0x5c475a60      4       1724    0x9496f7c6      2017-08-04 19:51:13
0x5c475d48      516     1720    0x77667098      2017-08-04 19:51:13     2017-08-04 19:52:35
0x5c478030      1072    1784    0x8aa26918      2017-08-04 19:51:13
0x5c480a88      1772    1776    0x77667098      2017-08-04 19:51:13
0x5c481d48      516     1740    0x77667098      2017-08-04 19:51:13     2012-06-04 20:01:29
0x5c48ad48      308     324     0x77667098      2012-06-01 19:57:01
0x5c48b9f0      880     1800    0x77667098      2017-08-04 19:51:13
0x5c48d3d8      1072    1744    0x8c8d3608      2017-08-04 19:51:13
0x5c48d9b8      424     2860    0x77667098      2012-06-04 20:00:40
0x5c48e200      1772    1792    0x8c9cd1d0      2017-08-04 19:51:13
0x5c48e4e8      1772    1788    0x77667098      2017-08-04 19:51:13
0x5c493cc8      1772    1812    0x77667098      2017-08-04 19:51:13
0x5c4944d8      1772    1820    0x77667098      2017-08-04 19:51:13
0x5c494978      2064    3904    0x77667098      2017-08-04 19:56:18
0x5c495bc8      516     2320    0x77667098      2012-06-04 20:00:50
0x5c496030      620     3628    0x77667098      2017-08-04 19:59:50
0x5c497b20      3432    2108    0x77667098      2012-06-04 20:01:05
0x5c49db90      1392    1852    0x91c47b90      2017-08-04 19:51:28
0x5c49e030      912     2972    0x77667098      2017-08-04 19:53:19
0x5c49f030      912     2964    0x77667098      2017-08-04 19:53:19
0x5c4a9a30      1072    1872    0x77667098      2017-08-04 19:51:28
0x5c4b0800      620     3296    0x77667098      2017-08-04 19:59:50
0x5c4b2d48      2012    2016    0x77667098      2017-08-04 19:51:28
0x5c4b3030      1176    1948    0x91c617f0      2017-08-04 19:51:28
0x5c4b6cb8      3064    724     0x77667098      2012-06-04 20:00:40
0x5c4b79c8      3064    3196    0x77667098      2017-08-04 19:53:31
0x5c4b7d48      2012    392     0x930ece40      2017-08-04 19:51:28
0x5c4bc230      2012    276     0x77667098      2017-08-04 19:51:28
0x5c4c13f0      2012    604     0x77667098      2017-08-04 19:51:28
0x5c4c2d48      3944    2756    0x77667098      2012-06-04 20:01:20
0x5c4c6a48      808     3268    0x77667098      2017-08-04 19:58:19
0x5c4c91f8      2064    2932    0x77667098      2012-06-01 19:57:42
0x5c4cb5e8      2012    280     0x77667098      2017-08-04 19:51:28
0x5c4cc0f8      2012    104     0x77667098      2017-08-04 19:51:28
0x5c4d1168      2012    2020    0x77667098      2017-08-04 19:51:28
0x5c4da030      912     3996    0x77667098      2017-08-04 19:56:20     2012-06-01 19:57:35
0x5c4dad18      4       2396    0x8c25ad18      2017-08-04 19:51:38
0x5c4dd918      292     296     0x77667098      2017-08-04 19:51:28
0x5c4de030      516     3272    0x77667098      2017-08-04 19:58:19
0x5c4ded48      2012    2028    0x77667098      2017-08-04 19:51:28
0x5c4e16e0      2012    264     0x91c8e610      2017-08-04 19:51:28
0x5c4e1ae0      2012    268     0x77667098      2017-08-04 19:51:28
0x5c4e2030      516     2180    0x930e29a8      2012-06-04 20:00:50
0x5c4e4608      2012    284     0x77667098      2017-08-04 19:51:28
0x5c4ec030      424     1276    0x77667098      2017-08-04 19:51:34
0x5c4f1030      292     592     0x77667098      2017-08-04 19:51:28
0x5c4f6030      292     956     0x77667098      2017-08-04 19:51:28
0x5c4f6728      2236    1676    0x77667098      2012-06-04 20:01:10
0x5c4f6a10      4       700     0x82a669c2      2017-08-04 19:51:28
0x5c4f6d48      292     656     0x8aa6a600      2017-08-04 19:51:28
0x5c4f8818      2012    1252    0x77667098      2017-08-04 19:51:29
0x5c4fa030      2012    1144    0x77667098      2017-08-04 19:51:28
0x5c4fa470      2244    1188    0x77667098      2012-06-01 19:57:23     2012-06-01 19:57:24
0x5c4fa758      292     1132    0x77667098      2017-08-04 19:51:28
0x5c4fb2d0      292     1124    0x77667098      2017-08-04 19:51:28
0x5c4fc8e0      292     904     0x77667098      2017-08-04 19:51:28
0x5c4fcd48      292     784     0x8aa39498      2017-08-04 19:51:28
0x5c4fd890      292     1092    0x77667098      2017-08-04 19:51:28
0x5c4fdcc8      292     1028    0x77667098      2017-08-04 19:51:28
0x5c4ff678      292     980     0x77667098      2017-08-04 19:51:28
0x5c50dd48      2064    3960    0x93831878      2017-08-04 19:56:20
0x5c5128e8      1308    3956    0x77667098      2012-06-04 20:00:50
0x5c5237a0      516     1224    0x77667098      2012-06-01 19:56:52
0x5c537030      424     1136    0x77667098      2017-08-04 19:51:34
0x5c537530      912     368     0x77667098      2017-08-04 19:51:34     2017-08-04 19:51:34
0x5c548a60      408     1300    0x77667098      2017-08-04 19:51:34
0x5c548d48      424     420     0x77667098      2017-08-04 19:51:34     2017-08-04 19:51:34
0x5c54a6d0      424     1844    0x77667098      2017-08-04 19:51:34
0x5c54bab0      396     412     0x77667098      2017-08-04 19:51:34
0x5c54c5a0      1072    352     0x77667098      2017-08-04 19:51:34     2017-08-04 19:53:34
0x5c54d030      2064    3328    0x93902850      2017-08-04 20:00:01
0x5c54f030      2176    2184    0x77667098      2012-06-04 20:00:50     2012-06-04 20:00:50
0x5c54f448      3944    952     0x77667098      2012-06-04 20:00:50
0x5c54fa58      2460    2464    0x882f45e0      2017-08-04 19:51:41
0x5c550d48      3432    3564    0x77667098      2012-06-04 20:00:41     2012-06-04 20:01:05
0x5c5533a8      292     3228    0x77667098      2017-08-04 20:00:08
0x5c559578      424     340     0x77667098      2017-08-04 19:51:34
0x5c55b680      396     1284    0x77667098      2017-08-04 19:51:34
0x5c55c030      396     2052    0x77667098      2017-08-04 19:51:34
0x5c55cd48      2064    2068    0x77667098      2017-08-04 19:51:34
0x5c578030      2064    2072    0x77667098      2017-08-04 19:51:35
0x5c578860      396     2076    0x77667098      2017-08-04 19:51:35
0x5c57dd48      396     2080    0x77667098      2017-08-04 19:51:35
0x5c5835f0      1360    2092    0x77667098      2017-08-04 19:51:35
0x5c583a58      424     2088    0x77667098      2017-08-04 19:51:35     2017-08-04 19:51:35
0x5c58ad48      424     2116    0x941d5448      2017-08-04 19:51:35
0x5c58c030      1360    2124    0x77667098      2017-08-04 19:51:35
0x5c58d768      1360    2128    0x77667098      2017-08-04 19:51:35
0x5c58f030      1360    2132    0x77667098      2017-08-04 19:51:35
0x5c58f898      1360    2136    0x88e27830      2017-08-04 19:51:35
0x5c591a18      1360    2144    0x77667098      2017-08-04 19:51:35
0x5c592030      2064    2148    0x85b92244      2017-08-04 19:51:35
0x5c593808      2064    2160    0x77667098      2017-08-04 19:51:35
0x5c593af0      2064    2156    0x77667098      2017-08-04 19:51:35
0x5c597298      1360    2164    0x77667098      2017-08-04 19:51:35
0x5c599618      2996    3256    0x77667098      2017-08-04 19:53:58
0x5c5a4d48      2064    2192    0x77667098      2017-08-04 19:51:35
0x5c5b33a8      2064    2208    0x77667098      2017-08-04 19:51:35
0x5c5b4cb8      2064    2216    0x77667098      2017-08-04 19:51:35
0x5c5b6348      2236    2240    0x82046760      2017-08-04 19:51:35
0x5c5b8d48      2064    3876    0x77667098      2017-08-04 19:56:17
0x5c5b9238      308     260     0x77667098      2012-06-01 19:57:01
0x5c5bcb18      2244    2248    0x91ce9b70      2017-08-04 19:51:35
0x5c5be628      2064    1196    0x77667098      2012-06-01 19:57:00
0x5c5c2d48      2244    2256    0x77667098      2017-08-04 19:51:35
0x5c5c3d48      2236    2260    0x77667098      2017-08-04 19:51:35
0x5c5cc030      2064    2264    0x77667098      2017-08-04 19:51:35
0x5c5ceb20      2460    3404    0x77667098      2017-08-04 19:58:27
0x5c5cfc70      2244    2276    0x77667098      2017-08-04 19:51:35
0x5c5d09f0      2064    2280    0x77667098      2017-08-04 19:51:35
0x5c5d3b20      2244    2288    0x77667098      2017-08-04 19:51:35
0x5c5d7a68      1176    2360    0x77667098      2017-08-04 19:51:36
0x5c5ddbe0      1392    2332    0x77667098      2017-08-04 19:51:35
0x5c5e10c0      2764    1664    0x77667098      2012-06-04 20:01:00
0x5c5e3258      2064    2508    0x77667098      2012-06-01 19:57:00
0x5c5e4860      1176    4036    0x77667098      2017-08-04 19:56:21
0x5c5e52f8      468     1504    0x77667098      2017-08-04 19:59:32
0x5c5ecd48      3064    3200    0x77667098      2017-08-04 19:53:31     2017-08-04 19:53:31
0x5c5ee3c0      2064    2492    0x77667098      2012-06-01 19:57:00
0x5c5f0d20      400     1496    0x77667098      2017-08-04 19:59:32
0x5c5f87f8      3064    4084    0x77667098      2017-08-04 19:56:27
0x5c600238      808     864     0x88e01858      2017-08-04 19:51:12
0x5c601b18      400     572     0x88ed4c58      2017-08-04 19:51:11
0x5c606828      912     1068    0x77667098      2017-08-04 19:51:12
0x5c607d48      3064    3084    0x77667098      2017-08-04 19:53:28
0x5c60a5b8      640     680     0x8f179990      2017-08-04 19:51:12
0x5c60ac78      640     672     0x77667098      2017-08-04 19:51:12
0x5c60b1b0      640     684     0x77667098      2017-08-04 19:51:12
0x5c6162d0      516     584     0x77667098      2017-08-04 19:51:12
0x5c618030      532     596     0x94150668      2017-08-04 19:51:12
0x5c61dbe8      516     612     0x77667098      2017-08-04 19:51:12     2017-08-04 19:54:36
0x5c6268a0      540     616     0x77667098      2017-08-04 19:51:12
0x5c62a030      640     648     0x77667098      2017-08-04 19:51:12
0x5c62aca0      640     652     0x77667098      2017-08-04 19:51:12     2017-08-04 19:52:48
0x5c62bd48      640     644     0x77667098      2017-08-04 19:51:12
0x5c62d1a8      516     692     0x8c908de0      2017-08-04 19:51:12
0x5c62dc20      640     688     0x77667098      2017-08-04 19:51:12
0x5c6386e8      640     704     0x8f1b9d08      2017-08-04 19:51:12
0x5c63f4b8      1308    1156    0x77667098      2012-06-04 20:01:20
0x5c641508      4       696     0x87848740      2017-08-04 19:51:12
0x5c646d48      880     1060    0x77667098      2017-08-04 19:51:12
0x5c649030      640     708     0x96c976e8      2017-08-04 19:51:12
0x5c64b7d0      912     2980    0x77667098      2017-08-04 19:53:19
0x5c64bbb0      712     716     0x77667098      2017-08-04 19:51:12
0x5c64e030      712     728     0x8584e244      2017-08-04 19:51:12     2017-08-04 19:54:36
0x5c6528e8      2460    1928    0x77667098      2017-08-04 19:59:46
0x5c653d48      712     740     0x77667098      2017-08-04 19:51:12
0x5c654938      712     744     0x88e23290      2017-08-04 19:51:12
0x5c657ab8      712     748     0x77667098      2017-08-04 19:51:12
0x5c65b030      1308    3928    0x77667098      2012-06-04 20:00:50
0x5c6614e8      540     756     0x77667098      2017-08-04 19:51:12
0x5c6618d8      540     752     0x77667098      2017-08-04 19:51:12
0x5c6661f0      540     760     0x77667098      2017-08-04 19:51:12
0x5c6683b8      540     772     0x77667098      2017-08-04 19:51:12     2017-08-04 19:54:36
0x5c6686a0      540     768     0x77667098      2017-08-04 19:51:12
0x5c668d48      540     764     0x77667098      2017-08-04 19:51:12
0x5c6692c0      540     776     0x77667098      2017-08-04 19:51:12
0x5c66dd48      4       788     0x82a669c2      2017-08-04 19:51:12
0x5c676d48      4       796     0x82a669c2      2017-08-04 19:51:12
0x5c678d48      712     3572    0x930e23e0      2017-08-04 19:59:47
0x5c682918      808     812     0x77667098      2017-08-04 19:51:12
0x5c684d48      4       824     0x82a669c2      2017-08-04 19:51:12
0x5c68ad48      4       828     0x82a669c2      2017-08-04 19:51:12
0x5c692d48      4       832     0x82a669c2      2017-08-04 19:51:12
0x5c697d48      808     840     0x77667098      2017-08-04 19:51:12
0x5c6b64c8      880     888     0x77667098      2017-08-04 19:51:12
0x5c6b7d48      880     884     0x77667098      2017-08-04 19:51:12
0x5c6b9c10      808     892     0x77667098      2017-08-04 19:51:12     2017-08-04 19:51:12
0x5c6ba880      880     900     0x94019888      2017-08-04 19:51:12
0x5c6bc9e0      808     924     0x77667098      2017-08-04 19:51:12
0x5c6bccc8      808     920     0x77667098      2017-08-04 19:51:12
0x5c6bd918      912     916     0x77667098      2017-08-04 19:51:12
0x5c6be930      808     928     0x77667098      2017-08-04 19:51:12
0x5c6c1030      912     932     0x77667098      2017-08-04 19:51:12
0x5c6c23e8      912     936     0x77667098      2017-08-04 19:51:12     2017-08-04 19:52:50
0x5c6c3030      912     3576    0x77667098      2017-08-04 19:59:50
0x5c6c3bf0      2244    2456    0x77667098      2012-06-01 19:57:23
0x5c6c6030      308     600     0x77667098      2012-06-01 19:57:01
0x5c6c8ad0      880     960     0x77667098      2017-08-04 19:51:12
0x5c6ca220      880     964     0x77667098      2017-08-04 19:51:12
0x5c6cdd48      880     972     0x77667098      2017-08-04 19:51:12
0x5c6d0988      880     1016    0x8c8344e0      2017-08-04 19:51:12
0x5c6d1030      808     988     0x77667098      2017-08-04 19:51:12
0x5c6e1030      808     1004    0x77667098      2017-08-04 19:51:12
0x5c6e2030      808     1008    0x8f189608      2017-08-04 19:51:12
0x5c6e5818      912     1032    0x77667098      2017-08-04 19:51:12
0x5c6e6030      912     1036    0x77667098      2017-08-04 19:51:12
0x5c6ea030      880     1048    0x8c899738      2017-08-04 19:51:12
0x5c6f7360      4       1052    0x8c2dc268      2017-08-04 19:51:12
0x5c6f8030      4       1056    0x8c2dc268      2017-08-04 19:51:12
0x5c6f8d48      880     1064    0x77667098      2017-08-04 19:51:12
0x5c6fa2f8      1072    1080    0x77667098      2017-08-04 19:51:12
0x5c6fbc68      1072    1076    0x77667098      2017-08-04 19:51:12
0x5c7033a0      516     2188    0x77667098      2012-06-04 20:00:50
0x5c7043b8      912     2660    0x77667098      2012-06-01 19:57:55
0x5c704a60      880     1100    0x77667098      2017-08-04 19:51:12
0x5c70c030      912     3280    0x77667098      2017-08-04 19:54:15
0x5c70d2f8      1072    1128    0x883bc5a8      2017-08-04 19:51:12
0x5c70dd48      880     1120    0x8c884c70      2017-08-04 19:51:12
0x5c70f7c0      880     1140    0x77667098      2017-08-04 19:51:12
0x5c70faa8      4       1864    0x91044005      2017-08-04 19:51:32
0x5c7127f0      620     2832    0x77667098      2017-08-04 19:59:50
0x5c718d48      808     1152    0x96c56c10      2017-08-04 19:51:12
0x5c71ba60      808     1172    0x77667098      2017-08-04 19:51:12
0x5c71c520      1176    1180    0x77667098      2017-08-04 19:51:12
0x5c728030      912     1240    0x77667098      2017-08-04 19:51:12
0x5c728a18      1176    1184    0x77667098      2017-08-04 19:51:12
0x5c72a370      1176    1192    0x77667098      2017-08-04 19:51:12     2012-06-01 19:57:27
0x5c72b380      808     1200    0x77667098      2017-08-04 19:51:12
0x5c730030      808     1204    0x77667098      2017-08-04 19:51:12
0x5c730a80      1176    1208    0x8c8d0398      2017-08-04 19:51:12
0x5c734030      1176    1220    0x77667098      2017-08-04 19:51:12
0x5c734430      1072    2992    0x77667098      2017-08-04 19:53:27     2017-08-04 19:56:10
0x5c735848      1176    1228    0x8f1aafe0      2017-08-04 19:51:12
0x5c7413e8      1176    1236    0x77667098      2017-08-04 19:51:12
0x5c7435f0      2064    288     0x77667098      2012-06-01 19:57:42
0x5c748030      912     1248    0x8c936e38      2017-08-04 19:51:12
0x5c750768      4       1892    0x8c25ad18      2017-08-04 19:51:32
0x5c751a20      3064    3092    0x85951c34      2017-08-04 19:53:28     2017-08-04 19:53:31
0x5c754478      912     1304    0x77667098      2017-08-04 19:51:12     2017-08-04 19:51:12
0x5c7547c0      2064    2432    0x82155468      2017-08-04 19:51:40
0x5c759d48      912     1312    0x77667098      2017-08-04 19:51:12
0x5c75a030      912     1316    0x8c924d30      2017-08-04 19:51:12
0x5c75b6a8      912     1320    0x77667098      2017-08-04 19:51:12
0x5c75e030      912     1324    0x77667098      2017-08-04 19:51:12
0x5c761d48      912     1328    0x77667098      2017-08-04 19:51:12     2017-08-04 19:52:35
0x5c763d48      2764    3800    0x77667098      2017-08-04 19:56:17
0x5c765200      912     3980    0x85965414      2017-08-04 19:56:20     2012-06-04 20:00:56
0x5c7691c8      4       1352    0x910446cb      2017-08-04 19:51:12
0x5c7694d0      4       1348    0x91044005      2017-08-04 19:51:12
0x5c7697f8      4       1344    0x91044005      2017-08-04 19:51:12
0x5c76a778      520     1360    0x77bb7098      2017-08-04 19:48:16
0x5c77b030      1360    1364    0x77667098      2017-08-04 19:51:12
0x5c781860      1360    1376    0x77667098      2017-08-04 19:51:12
0x5c7829c0      1360    3772    0x77667098      2017-08-04 19:56:17
0x5c783d48      1360    1384    0x77667098      2017-08-04 19:51:12
0x5c786a30      1360    1388    0x91c39390      2017-08-04 19:51:12
0x5c78a030      1392    1396    0x77667098      2017-08-04 19:51:12
0x5c78b468      1392    1400    0x77667098      2017-08-04 19:51:12
0x5c78c3f0      2460    3592    0x77667098      2017-08-04 19:54:47     2017-08-04 19:54:47
0x5c798cc8      1392    1416    0x77667098      2017-08-04 19:51:12
0x5c7a81d8      1392    1420    0x77667098      2017-08-04 19:51:13
0x5c7ae880      1392    1440    0x8c9147b8      2017-08-04 19:51:13
0x5c7afd48      4       1424    0x910c1fc0      2017-08-04 19:51:13
0x5c7b4030      1392    1432    0x77667098      2017-08-04 19:51:13
0x5c7b8030      1392    1436    0x859b8244      2017-08-04 19:51:13     2017-08-04 19:51:13
0x5c7b8498      1392    1444    0x859b86ac      2017-08-04 19:51:13     2012-06-04 20:01:32
0x5c7b9350      2064    2440    0x77667098      2017-08-04 19:51:40
0x5c7c18a0      2460    3708    0x77667098      2017-08-04 19:55:56     2012-06-04 20:00:48
0x5c7c2a40      408     2416    0x90434948      2017-08-04 19:51:38
0x5c7c2d28      408     2412    0x90436c14      2017-08-04 19:51:38
0x5c7c3558      4       2392    0x8c25ad18      2017-08-04 19:51:38
0x5c7c7d48      1392    1452    0x9402c530      2017-08-04 19:51:13
0x5c7cd980      540     3712    0x77667098      2017-08-04 20:00:14
0x5c7cea60      1176    1468    0x940163a8      2017-08-04 19:51:13
0x5c7ced48      1392    1456    0x8c92c9d8      2017-08-04 19:51:13
0x5c7d2030      1392    1460    0x77667098      2017-08-04 19:51:13
0x5c7d2a40      1392    1464    0x77667098      2017-08-04 19:51:13
0x5c7d9978      4       1476    0x911bfe2e      2017-08-04 19:51:13
0x5c7d9c70      4       1472    0x9117fe22      2017-08-04 19:51:13
0x5c7da788      1392    1484    0x77667098      2017-08-04 19:51:13
0x5c7eed48      1392    1492    0x77667098      2017-08-04 19:51:13
0x5c7fc758      1176    1500    0x77667098      2017-08-04 19:51:13     2017-08-04 19:53:19
0x5c7fe760      1072    1960    0x77667098      2017-08-04 19:51:28
0x5c999030      348     376     0x77667098      2017-08-04 19:51:11
0x5c999c90      348     380     0x8f171218      2017-08-04 19:51:11
0x5c99a288      348     388     0x77667098      2017-08-04 19:51:11
0x5c99a6b8      348     384     0x77667098      2017-08-04 19:51:11
0x5c9a01d8      348     416     0x77667098      2017-08-04 19:51:11
0x5c9a17d8      400     404     0x96c39bf0      2017-08-04 19:51:11
0x5c9a9d48      912     2252    0x77667098      2017-08-04 19:56:29
0x5c9b15d0      408     456     0x8aa42210      2017-08-04 19:51:11
0x5c9b1930      408     452     0x77667098      2017-08-04 19:51:11
0x5c9b4d48      4       428     0x82a669c2      2017-08-04 19:51:11
0x5c9b7d48      400     436     0x77667098      2017-08-04 19:51:11
0x5c9b9928      4       448     0x82e0a4e9      2017-08-04 19:48:15
0x5c9b9cc8      400     440     0x77667098      2017-08-04 19:51:11     2017-08-04 19:52:34
0x5c9bd4f0      636     672     0x77bb7098      2017-08-04 19:48:15     2017-08-04 19:50:45
0x5c9bdd48      4       448     0x82e0a4e9      2017-08-04 19:51:11
0x5c9c1350      640     660     0x77667098      2017-08-04 19:51:12
0x5c9c1d08      640     664     0x77667098      2017-08-04 19:51:12     2017-08-04 19:51:12
0x5c9c3590      408     460     0x77667098      2017-08-04 19:51:11
0x5c9c4030      408     464     0x77667098      2017-08-04 19:51:11
0x5c9c52b8      408     476     0x77667098      2017-08-04 19:51:11
0x5c9c5940      468     472     0x96c39a48      2017-08-04 19:51:11
0x5c9cc190      516     2196    0x77667098      2012-06-04 20:00:50     2012-06-04 20:00:50
0x5c9d1d48      2236    1976    0x77667098      2017-08-04 19:59:41
0x5c9d27f8      808     860     0x857d2a0c      2017-08-04 19:51:12     2017-08-04 19:51:12
0x5c9d3340      468     484     0x77667098      2017-08-04 19:51:11
0x5c9d3d48      808     676     0x77667098      2017-08-04 20:00:01
0x5c9d4b70      808     856     0x77667098      2017-08-04 19:51:12     2017-08-04 19:52:34
0x5c9d8d48      408     496     0x77667098      2017-08-04 19:51:11
0x5c9d9768      348     504     0x77667098      2017-08-04 19:51:11
0x5c9d9a50      348     500     0x77667098      2017-08-04 19:51:11
0x5c9e5030      516     520     0x857e5244      2017-08-04 19:51:11     2017-08-04 19:51:58
0x5c9e5850      348     528     0x77667098      2017-08-04 19:51:11
0x5c9e66d0      4       3192    0x82a669c2      2017-08-04 19:53:31
0x5c9e6d48      912     3164    0x77667098      2017-08-04 19:53:29     2017-08-04 19:58:35
0x5c9eba70      400     524     0x77667098      2017-08-04 19:51:11     2017-08-04 19:52:34
0x5c9ee440      540     544     0x88e20460      2017-08-04 19:51:11
0x5c9f0850      532     632     0x857f0a64      2017-08-04 19:51:12     2012-06-01 19:57:26
0x5c9f1d48      532     548     0x77667098      2017-08-04 19:51:11
0x5c9f4810      348     552     0x77667098      2017-08-04 19:51:11
0x5c9f50c8      532     556     0x77667098      2017-08-04 19:51:11
0x5c9fc668      808     868     0x88e4e230      2017-08-04 19:51:12
0x5c9fd3f0      532     564     0x88e9ecd8      2017-08-04 19:51:11
0x5c9fd978      532     560     0x77667098      2017-08-04 19:51:11
0x5c9feae0      532     568     0x91c45748      2017-08-04 19:51:11
0x5cada320      408     508     0x88ea66e8      2017-08-04 19:51:11
0x5cae29b0      4       336     0x8f623e32      2017-08-04 19:51:11
0x5cb0aad0      4       372     0x8ef13406      2017-08-04 19:51:11
0x5cc91d48      4       172     0x8710d26e      2017-08-04 19:51:06
0x5d00a718      4       236     0x878e2522      2017-08-04 19:51:10
0x5d02cd48      4       240     0x8c249bcb      2017-08-04 19:51:10
0x5d037438      4       244     0x8c327e8a      2017-08-04 19:51:10
0x5d03b728      2064    2452    0x77667098      2017-08-04 19:51:41
0x5d03bd48      808     4004    0x77667098      2017-08-04 19:56:20
0x5d03f7a0      4       248     0x82a4301c      2017-08-04 19:51:10
0x5d043808      252     256     0x77667098      2017-08-04 19:51:10
0x5d04e788      516     628     0x77667098      2017-08-04 19:51:12
0x5d06ed48      308     3964    0x77667098      2012-06-04 20:00:40
0x5d072888      4       300     0x8ee8f52e      2017-08-04 19:51:10
0x5d099d48      4       304     0x8ee7de0d      2017-08-04 19:51:10
0x5d0c22c0      252     312     0x77667098      2017-08-04 19:51:10
0x5d0cc900      1308    3948    0x77667098      2012-06-04 20:00:50
0x5d0d0d48      516     624     0x77667098      2012-06-04 20:01:30
0x5d0d3460      4       328     0x8c38f005      2017-08-04 19:51:10
0x5d278240      4       192     0x874ad29f      2017-08-04 19:51:10
0x5d2788e8      4       188     0x874b7fea      2017-08-04 19:51:10
0x5d278bc0      4       184     0x874c891b      2017-08-04 19:51:10
0x5d2d42b0      4       200     0x877b6eb0      2017-08-04 19:51:10
0x5d2d4628      4       196     0x877b6eb0      2017-08-04 19:51:10
0x5d2d5020      4       204     0x877b6eb0      2017-08-04 19:51:10
0x5d2d5658      4       216     0x877b6eb0      2017-08-04 19:51:10
0x5d2d59d0      4       212     0x877b6eb0      2017-08-04 19:51:10
0x5d2d5d48      4       208     0x877b6eb0      2017-08-04 19:51:10
0x5d2d7020      4       220     0x877b6eb0      2017-08-04 19:51:10
0x5d2d7a70      4       228     0x877b6eb0      2017-08-04 19:51:10
0x5d2d7d48      4       224     0x877b6eb0      2017-08-04 19:51:10
0x5d368228      2244    2656    0x77667098      2012-06-04 20:00:59
0x5d36a030      712     1956    0x77667098      2017-08-04 20:00:14
0x5d60a350      912     608     0x77667098      2012-06-01 19:57:01
0x5d60b9a8      308     780     0x77667098      2012-06-01 19:57:01
0x5d6161d0      2064    2448    0x77667098      2012-06-01 19:57:00
0x5d6178a8      912     2840    0x77667098      2012-06-04 20:00:35
0x5d617b90      2064    2716    0x77667098      2012-06-01 19:57:01
0x5d629030      3944    2316    0x77667098      2012-06-04 20:00:50
0x5d6293b0      3944    2300    0x77667098      2012-06-04 20:00:50     2012-06-04 20:00:50
0x5d62d8d0      540     1964    0x77667098      2017-08-04 20:00:14
0x5d636030      2064    1680    0x77667098      2012-06-01 19:57:00
0x5d637d20      2064    636     0x77667098      2012-06-01 19:57:02
0x5d641d48      880     1848    0x77667098      2012-06-04 20:00:51     2012-06-04 20:00:51
0x5d895340      2764    2772    0x77667098      2017-08-04 19:52:28
0x5d895628      2764    2768    0x77667098      2017-08-04 19:52:28
0x5d895a58      2764    2776    0x77667098      2017-08-04 19:52:28     2017-08-04 19:53:19
0x5d89c8e0      912     232     0x77667098      2017-08-04 19:51:34     2017-08-04 19:51:34
0x5d89d780      2064    2436    0x8469d994      2017-08-04 19:51:40
0x5d89f030      424     1292    0x77667098      2017-08-04 19:51:34
0x5d8af030      620     3300    0x77667098      2017-08-04 19:59:50
0x5d8b0030      516     2908    0x77667098      2012-06-04 20:00:22
0x5d8b2c20      712     2560    0x77667098      2012-06-01 19:57:27
0x5d8b33b8      880     3684    0x77667098      2017-08-04 20:00:01
0x5d8f0030      712     2384    0x77667098      2012-06-04 20:00:50
0x5d8f0490      1392    3072    0x77667098      2012-06-04 20:00:57
0x5d8f0778      3944    2988    0x93897470      2012-06-04 20:00:50
0x5d8f0d48      3944    316     0x77667098      2012-06-04 20:00:50
0x5d8f1940      3432    3976    0x77667098      2012-06-04 20:00:41     2012-06-04 20:01:05
0x5d8f1940      3432    3976    0x77667098      2012-06-04 20:00:41     2012-06-04 20:01:05
0x5d8f1940      3432    3976    0x77667098      2012-06-04 20:00:41     2012-06-04 20:01:05
0x5d91b750      4       176     0x871d319c      2017-08-04 19:51:07
0x5d91c750      4       180     0x871d7848      2017-08-04 19:51:07
0x5d929a60      880     2084    0x938c7500      2012-06-04 20:00:41
0x5d929d48      3064    3088    0x9401e9e0      2017-08-04 19:53:28
0x5d935030      3064    3112    0x77667098      2017-08-04 19:53:28
0x5d935cc8      3064    3116    0x77667098      2017-08-04 19:53:28
0x5d936030      3064    3068    0x77667098      2017-08-04 19:53:28
0x5d937d48      912     3136    0x77667098      2017-08-04 19:53:29     2017-08-04 19:58:35
0x5d9a0718      620     536     0x93149ce0      2017-08-04 19:59:50
0x5d9a0a00      2064    1856    0x77667098      2012-06-01 19:57:02
0x5d9a2d48      400     492     0x77667098      2012-06-04 20:01:31
0x5d9a3c80      2064    3212    0x77667098      2012-06-01 19:57:42
0x5d9a4b10      424     3148    0x77667098      2012-06-01 19:57:24
0x5d9a7858      1116    1580    0x77667098      2012-06-04 20:00:51     2012-06-04 20:01:05
0x5d9a8d48      3944    3952    0x77667098      2012-06-04 20:00:50
0x5d9add48      4       3060    0x93266f2e      2017-08-04 19:53:28     2017-08-04 19:58:35
0x5d9aeb48      808     3460    0x77667098      2012-06-04 20:00:40
0x5d9b0338      3944    2356    0x77667098      2012-06-04 20:00:50
0x5d9b2b38      2764    1668    0x77667098      2012-06-04 20:01:00
0x5d9b4510      292     588     0x77667098      2012-06-04 20:01:20
0x5d9b4d48      712     2404    0x77667098      2012-06-04 20:00:50
0x5d9b6338      3064    3108    0x77667098      2017-08-04 19:53:28
0x5d9b6620      3064    3104    0x77667098      2017-08-04 19:53:28
0x5d9bad48      1072    3324    0x77667098      2017-08-04 20:00:01
0x5d9bb688      808     3124    0x930e2608      2017-08-04 19:53:29
0x5d9bc030      912     3128    0x77667098      2017-08-04 19:53:29
0x5d9bd850      2064    1588    0x77667098      2012-06-01 19:57:00
0x5d9be030      808     3132    0x9317a278      2017-08-04 19:53:29
0x5d9c3588      308     3852    0x77667098      2012-06-04 20:00:40
0x5d9ccd48      640     1908    0x77667098      2017-08-04 19:59:46
0x5d9d7328      3064    3172    0x77667098      2017-08-04 19:53:30
0x5d9d8ba8      1072    3548    0x77667098      2017-08-04 19:54:30     2017-08-04 19:56:30
0x5d9da4a8      3064    3508    0x930bb810      2017-08-04 19:54:28
0x5d9db498      620     3304    0x77667098      2017-08-04 19:59:50
0x5d9eba60      468     3872    0x77667098      2012-06-04 20:00:40
0x5d9ebd48      468     2704    0x77667098      2012-06-01 19:57:25
0x5d9ef750      712     2312    0x77667098      2012-06-04 20:00:50     2012-06-04 20:00:50
0x5d9f0d48      1116    1756    0x77667098      2012-06-04 20:00:51     2012-06-04 20:01:05
0x5d9f0d48      1116    1756    0x77667098      2012-06-04 20:00:51     2012-06-04 20:01:05
0x5d9f5ce0      1116    836     0x77667098      2012-06-04 20:00:51     2012-06-04 20:01:05
0x5d9f59f8      1116    2720    0x77667098      2012-06-04 20:00:51     2012-06-04 20:01:05
0x5d9f5ce0      1116    836     0x77667098      2012-06-04 20:00:51     2012-06-04 20:01:05
0x5d9f7180      1392    3420    0x77667098      2017-08-04 19:58:51
0x5d9f8938      3064    3516    0x77667098      2017-08-04 19:54:28
0x5d9fd508      2244    1924    0x77667098      2012-06-04 20:00:59
0x5d9ffa60      880     344     0x77667098      2012-06-04 20:00:51
0x5d9ffd48      880     2272    0x77667098      2012-06-04 20:00:51
0x5dab54c8      4       148     0x82a669c2      2017-08-04 19:51:05
0x5dab9a88      4       152     0x82a669c2      2017-08-04 19:51:05
0x5dabf7a0      2996    3004    0x77667098      2017-08-04 19:53:28
0x5dabfa88      2996    3000    0x77667098      2017-08-04 19:53:28
0x5dac1d48      2996    3008    0x77667098      2017-08-04 19:53:28
0x5dac2030      2996    3020    0x9419b3c0      2017-08-04 19:53:28
0x5dac2968      3432    3380    0x77667098      2012-06-04 20:00:41     2012-06-04 20:01:05
0x5dacf020      4       156     0x82a4f6ac      2017-08-04 19:51:05
0x5db38848      4       128     0x8286c537      2017-08-04 19:51:05
0x5db3a020      4       72      0x828bb99e      2017-08-04 19:51:05
0x5db3a4b0      4       84      0x828719ad      2017-08-04 19:51:05
0x5db3a788      4       80      0x828719ad      2017-08-04 19:51:05
0x5db3ad48      4       76      0x82a6f139      2017-08-04 19:51:05
0x5db3b020      4       52      0x828bb99e      2017-08-04 19:51:05
0x5db3b4c0      4       68      0x828bb99e      2017-08-04 19:51:05
0x5db3b798      4       64      0x828bb99e      2017-08-04 19:51:05
0x5db3ba70      4       60      0x828bb99e      2017-08-04 19:51:05
0x5db3bd48      4       56      0x828bb99e      2017-08-04 19:51:05
0x5db3c6e8      4       124     0x8283f550      2017-08-04 19:51:05
0x5db3c9c0      4       120     0x8283f550      2017-08-04 19:51:05
0x5db3d8e0      4       20      0x82853b1c
0x5db3dbb8      4       16      0x82853b1c
0x5db3f7d0      4       116     0x828d85ab      2017-08-04 19:51:05
0x5db416c0      4       12      0x82840a19
0x5db431d0      4       92      0x82870e9f      2017-08-04 19:51:05
0x5db43570      4       88      0x828a015e      2017-08-04 19:51:05
0x5db45798      4       32      0x828bb99e      2017-08-04 19:51:05
0x5db45a70      4       28      0x828bb99e      2017-08-04 19:51:05
0x5db45d48      4       24      0x828bb99e      2017-08-04 19:51:05
0x5db474c0      4       108     0x828b3973      2017-08-04 19:51:05
0x5db47798      4       168     0x870bb95e      2017-08-04 19:51:06
0x5db47a70      4       100     0x828c782c      2017-08-04 19:51:05
0x5db47d48      4       96      0x828b6d80      2017-08-04 19:51:05
0x5db49020      4       36      0x828bb99e      2017-08-04 19:51:05
0x5db494e0      4       48      0x828bb99e      2017-08-04 19:51:05
0x5db497b8      4       44      0x828bb99e      2017-08-04 19:51:05
0x5db49a90      4       40      0x828bb99e      2017-08-04 19:51:05
0x5db55388      4       132     0x82a669c2      2017-08-04 19:51:05
0x5db59020      4       160     0x870c0c80      2017-08-04 19:51:05
0x5db60c90      2064    2668    0x94044930      2017-08-04 19:51:45
0x5db61448      2764    2796    0x77667098      2017-08-04 19:52:28     2012-06-01 19:57:27
0x5db61730      2764    2792    0x77667098      2017-08-04 19:52:28
0x5db62d48      4       136     0x82a669c2      2017-08-04 19:51:05
0x5db6d370      4       144     0x82a669c2      2017-08-04 19:51:05
0x5db6d908      4       140     0x82a669c2      2017-08-04 19:51:05
0x5dbb7868      4       8       0x829cb433

C:\volatility3>
```

Please review this pull request and consider merging it into the main branch. Your feedback and suggestions for improvement are welcome. Thank you!

RuBublik committed:
added 'PoolConstraint' of Thread objects to 'PoolScanner.default_constraints' as part of adding support for thread pool tag scanning

RuBublik committed:
temporary fix to ETHREAD class, add 'is_valid' method

RuBublik committed:
added permanent implementation for 'ETHREAD.is_valid'

RuBublik committed:
twicked thread constraint for edge cases

RuBublik committed:
fix of is_valid - removed reliace on owning _eprocess and added exclusion for system process (does not have creation time)

RuBublik committed:
added thrdscan plugin to utilize added support for ethread pool tag scanning

RuBublik committed:
added test for thrdscan plugin

ikelos commented on test/test_volatility.py:
```diff
@@ -189,6 +189,16 @@ def test_windows_svcscan(image, volatility, python):
 
 
+def test_windows_thrdscan(image, volatility, python):
+    rc, out, err = runvol_plugin("windows.thrdscan.ThrdScan", image, volatility, python)
```
Wow!  This is awesome!  Thank you for adding the tests in already!  Extra bonus marks!!!  5:D

ikelos commented on volatility3/framework/plugins/windows/thrdscan.py:
```diff
@@ -0,0 +1,107 @@
+    """Scans for windows threads."""
+
+    # cuz installed Framework interface version 2
+    _required_framework_version = (2, 0, 0)     
```
Since you've added a `scan_threads` classmethod, suggesting you want the plugin's functionality to be used by other plugins, this should really define `_version` as well, so that we can make changes to it in the future and plugins will clearly break if they're used against an older version that doesn't have a needed feature...  5:)

ikelos commented on volatility3/framework/plugins/windows/thrdscan.py:
```diff
@@ -0,0 +1,107 @@
+        layer_name: str,
+        symbol_table: str,
+    ) -> Iterable[interfaces.objects.ObjectInterface]:
+        """Scans for threads using the poolscanner module and constraints.
```
Even more extra bonus points for docstrings!!!  For a first time submission, you're nailing it!  5:D

ikelos commented on volatility3/framework/plugins/windows/thrdscan.py:
```diff
@@ -0,0 +1,107 @@
+                owner_proc_pid,
+                thread_tid,
+                hex(thread_start_addr),
+                str(thread_create_time) if isinstance(thread_create_time, datetime.datetime) else "",
```
So, the `datetime.datetime` is actually an acceptable simpletype as seen by volatility, and it makes sense to keep them that way so that user interfaces can sort them or make display adjustments to how the time is displayed and so on.  Also, if the object isn't a datetime, then you should return an object that inherits from `BaseAbsentValue` which will displays as `-` in the standard CLI, but will allow other user interfaces to decide how best to display that information.  5:)

ikelos commented on volatility3/framework/plugins/windows/thrdscan.py:
```diff
@@ -0,0 +1,107 @@
+                ("PID", int),
+                ("TID", int),
+                ("Start Address", str),
+                ("Create Time", str),
```
These should then be change from `str` to `datetime.datetime`.  5:)

Also, since you're dealing with times, you can get this plugin's output automatically added to the timeliner plugin by just implementing the TimelinerInterface](https://github.com/volatilityfoundation/volatility3/blob/8e56cb39d13c22575ba5133ffbc6b83344bed0ba/volatility3/framework/plugins/timeliner.py#L28) (remember to inherit from `volatility3.plugins.timeliner`, not `volatility3.plugins.framework.timeliner`).  5:)

ikelos commented on volatility3/framework/symbols/windows/extensions/__init__.py:
```diff
@@ -492,9 +492,39 @@ def get_name(self) -> str:
+            # passed all valitations
+            return True
+        except:
+            return False
```
When an exception occurs, it's best to log it (using `constants.LOGLEVEL_VVVV` if it's potentially quite noisy) just so there's a way to diagnose that somethign unexpected happened.  5:)

RuBublik replied:
regarding this -
I completely understand your concern. In general, you are absolutely right, and it's a good practice.
However, in this specific case, I noticed that other classes' implementations of `is_valid` do not log the exceptions, and that's why I did not log it either. In particular, I found this pattern in the `EPROCESS` and `CONTROL_AREA` classes.
Based on my understanding, exceptions that occur in that clause indicate that the object that's being checked is not a proper `_Ethread` kernel struct, but random data found after a byte sequence matching the pool tag. Considering the expected frequency of these exceptions, I decided to comply with the implementations of `is_valid` in the other two classes.
If you think that would be beneficial, I will of course add logging to the exception handler.
Please just let me know your preference (:

ikelos replied:
Yep, that's fair enough, I didn't spot the function it was in, and you're right, if someone's testing for validity, it's not critical they're told why it failed since they should end up still handling the issue gracefully.  Happy to mark this as resolved.

ikelos reviewed (commented):
Thanks very much, for a first submission you didn't trip up on any of the usual things that catch people out!  There were really only a couple of points (returning datetimes rather than strings and AbsentValues instead of empty strings), but great work with the docstrings and the tests!!!  5:D

There's also a pretty quick addition that would add this functionality to the timeliner plugin automatically, so please have a look at that and consider adding it (no problem if you don't want to, or just ask if you need a hand doing it).

It'll have to go for review by @iMHLv2 (our resident windows expert) just to catch any subtleties with thread scanning, and as a second pair of eyes, but hopefully we can get that done soon and get this merged.  Thanks again!  5:)

ikelos commented on volatility3/framework/symbols/windows/extensions/__init__.py:
```diff
@@ -492,9 +492,39 @@
+
+            # passed all valitations
+            return True
+        except:
```
Oh, yep, missed this, please also use specific exception handlers for what you're trying to catch.  Just using `except` without a type can grab all kinds of unexpected exceptions, that it's better to let bubble up to a higher exception handler).  5:)

ikelos commented:
Please also run black against this to ensure the code is suitably formatted for the project.  If you don't have black installed, you can get the diff of changes it would make by clicking on details for the black test with the red X by it.

RuBublik committed:
fixed typo

RuBublik committed:
changed TreeGrid yielded types to specific simpletypes instead of str

RuBublik committed:
implemented generate_timeline method in ThrdScan

RuBublik committed:
added _version to ThrdScan

RuBublik committed:
formated with black

RuBublik commented:
Thank you very much for your response and feedback! (: 
I have addressed the issues you mentioned and made the necessary corrections to the code. Here are the updates I made:
- The `thrdscan` plugin now correctly returns datetime and `format_hints.Hex` simpletypes.
- I have added the _version member as you suggested.
- Additionally, I ran the code through the *black* formatter for consistent formatting.
Furthermore, I have implemented the `timeliner` functionality you mentioned. 

I had only one question regarding logging in the exception handler in `ETHREAD.is_valid` method - 
I noticed that other classes' implementations of `is_valid` do not log the exceptions, and that's why I did not log it either. From what I understand, exceptions that occur in that clause simply mean that the object that's being checked is not a proper _Ethread kernel struct, but random data found after a byte sequence matching the pool tag, and these exceptions are expected to occur frequently.
If you think that would be beneficial, I will of course add logging to the exception handler. 
Please let me know your preference (:

Looking forward to your feedback and any additional suggestions, along with your opinion regarding the further question.
Thanks again! (:

ikelos commented:
Hiya, thanks very much for making those changes so quickly!  I'll give it another review when I have time, but yep your reasoning about the `is_valid` method seems reasonable, so happy to keep it as is.  For some reason black still seems to have some issues, I don't know if it's a different version of black or not, but if you can't figure it out, we can fix it up after merge.  Now we're just waiting on a quick check over by @iMHLv2 ...  5:)

digitalisx commented on volatility3/framework/plugins/windows/thrdscan.py:
```diff
@@ -0,0 +1,141 @@
+    """Scans for windows threads."""
+
+    # cuz installed Framework interface version 2
+    _required_framework_version = (2, 0, 0)
```
This PR changes the existing `_ETHREAD` API.. In other words, this plugin is only available after a change.
Then, I think that at the point of this merging, the version of the framework or the version of the requirement framework should be changed.

RuBublik replied:
Hey, thanks for your comment! I appreciate your observation regarding the changes made to the _ETHREAD API in this pull request. I understand your concern about the version of the framework and the requirements.

Could you please provide more information about the specific change you are suggesting in relation to the framework version or requirement framework? Since I'm not deeply familiar with the working and versioning conventions on this project, it will help me better understand your suggestion and make any necessary adjustments.

Thank you again for your feedback, and I appreciate your help! (:

ikelos replied:
Hi there, because this introduces new code to the core threads code (changes in the framework, not just to a plugin), any code that makes use of those changes (including yours) needs to be able to tell if they'll be there or not.  It looks like you've only made an additive change, so you only need to bump the minor version (in `constants/__init__.py`) and then set this plugin's `_required_framework_version` to the same value, so that if someone takes this plugin to another version of volatility, it will fail gracefully if the version is too old.  Otherwise it would work fine until it tried to make use of the new feature and throw an exception which the user would have no idea about.  5;)  I hope that explains the changes that would need to be made?

If you'd changed something, or removed something (so that an old plugin stopped working), then you'd need to change the major version number and we'd need to evaluate if it was a worthwhile/necessary change, since all older plugins would stop working and everything would need their version requirements updating.  Hope that clarifies about the version numbers a bit?  I'll see if I can get @iMHLv2 to have a review of this sometime soon...  5:)

RuBublik replied:
Hi ikelos,
Thank you for providing such an elaborate explanation. The modification that @digitalisx suggested is much clearer now.
I will certainly make the needed changes as soon as I can. In fact, I might be able to work on it this weekend. 
Appreciate your patience and guidance in this process (:

RuBublik replied:
Hi, 
I have made the necessary changes to the versions in the relevant places.
Bumped the `MINOR_VERSION` version from 4 to 5 (in `constants/__init__.py` file), and the `_required_framework_version` of the `thrdscan` plugin to 2.5.2, accordingly.

RuBublik committed:
fix black formatting

RuBublik commented:
Hey, Thanks for your response!
Regarding the issue with the black formatter, I've made some adjustments on my machine and reformatted the changed files again. Hopefully, it looks better now. 
Looking forward to your review @iMHLv2 (-:

iMHLv2 commented on volatility3/framework/plugins/windows/poolscanner.py:
```diff
@@ -222,6 +222,24 @@ def builtin_constraints(
+                b"Thr\xe5",  # -> “protected” allocation, MSB is set.
+                type_name=symbol_table + constants.BANG + "_ETHREAD",
+                object_type="Thread",
+                size=(600, None),  # -> 0x0258 - size of strcut in win5.1
```
Typo in `strcut` (struct?)

RuBublik replied:
You are right. Thanks for pointing this our to me. I will correct this right away.

RuBublik replied:
Hi iMHLv2,
I have corrected the typo you mentioned. 
Thx (:

iMHLv2 commented on volatility3/framework/plugins/windows/poolscanner.py:
```diff
@@ -222,6 +222,24 @@ def builtin_constraints(
+                b"Thre",
+                type_name=symbol_table + constants.BANG + "_ETHREAD",
+                object_type="Thread",
+                size=(600, None),  # -> 0x0258 - size of strcut in win5.1
```
Typo in `strcut` (struct?)

RuBublik replied:
You are right. Thanks for pointing this our to me. I will correct this right away.

RuBublik replied:
Hi iMHLv2,
I have corrected the typo you mentioned.
Thx (:

iMHLv2 commented on volatility3/framework/symbols/windows/extensions/__init__.py:
```diff
@@ -492,9 +492,41 @@ def get_name(self) -> str:
+        # passed all validations
+        return True
+
+    def get_create_time(self):
```
I noticed when testing this plugin that the creation timestamps were off, although the exit times were as expected for this sample:

```
0x1a03a30	1436	1112	0x7c8106e9	4891-04-14 17:32:21.000000 	2012-04-14 23:11:48.000000 
0x1a03da8	696	1244	0x7c8106e9	4891-04-14 21:24:12.000000 	2012-04-14 23:40:31.000000 
0x1a05da8	1096	464	0x7c8106e9	4891-04-14 21:24:12.000000 	2012-04-14 23:40:31.000000 
```

For some reason, these two timestamps are not in the same format. Check out how vol2 defines CreateTime as a "ThreadCreateTimeStamp" (https://github.com/volatilityfoundation/volatility/blob/master/volatility/plugins/overlays/windows/windows.py#L69). 

The underlying difference is that some bits are shifted for the creation timestamp: https://github.com/volatilityfoundation/volatility/blob/master/volatility/plugins/overlays/windows/windows.py#L1033

RuBublik replied:
Hi,
Thanks for bringing this serious issue to my attention. I will certainly look into it, and try to address the problem very soon. In fact, I might be able to work on it this weekend.
In the meantime, thx again for your response and feedback! (:

RuBublik replied:
Hello @iMHLv2.

After thoroughly looking into the thread creation timestamps and examining the implementation in the vol2 project, I tested on various win7 and win10 mem images (widely-known win7 sample 'stuxnet.vmem', as well as snapshots taken from local Windows virtual machines). 
Unfortunately, I was unable to replicate the issue you encountered, as the timestamps appeared to be accurate, and same with thrdacan of vol2 and my thrdscan for vol3. It appears that the existing implementation of the plugin is functioning correctly, however there may still be differences between the memory sample you tested and the ones I tested.

I would be happy if you could please provide the memory sample that has caused the problem, so I could test the plugin on it and further look into the issue.

Thank you for assistance and feedback!

RuBublik replied:
Hi @iMHLv2 and @ikelos,
Do you have any update on this? 
It would immensely help me to address your request if you could provide the memory sample that you encountered the bug with, or some guidance or instruction on how to set up my environment to get an equal dump, so I could reproduce the problem you mentioned with the plugin on my end.
Really looking forward to cooperate and contribute to the project!

iMHLv2 replied:
Hi @RuBublik - Sorry for the delay. I couldn't find the exact sample, but I do have several others with the same problem. It appears to be limited to XP samples. Here's an example to play with:

https://volexity-memory-sample-share.s3.amazonaws.com/grrcon.img.zip

RuBublik replied:
Hi @iMHLv2 ,
Thank you so much for your response and pinpointing of the problem!
I have tested the plugin against your sample, and indeed reproduced the problem. You were right about the bit shifting, it is present on older windows versions, like XP.
I have revised the code closely replicating the implementation in vol2 you mentioned (shifting the timestamps which `ETHREAD.get_create_time()` returns for older windows versions, when this is needed). 
After thoroughly testing this `thrdscan` plugin of vol3 in comparison to the counterpart in vol2 on several memory samples of various windows versions - I'm sure the timestamps are correct now.

Looking forward to your review!
Hope that we can proceed with merging this time (-:

RuBublik replied:
Hi @iMHLv2,
I've addressed the changes you requested in the pull request. Any chance you could take a moment to review them? 
Looking forward to your approval and merging the PR. 
Thanks!

iMHLv2 replied:
@RuBublik It looks good now, thanks and sorry for the delay.

ikelos commented:
@RuBublik any chance you could review the comments we've made please?  We can't move forward with this at the moment otherwise I'm afraid...

RuBublik committed:
bumped MINOR_VERSION to 2.5.2 after changes, and updated dependent thrdscan plugin's required version to this

RuBublik committed:
fixed typos

ikelos commented:
Thanks for making those changes, the BUILD number resets to 0 when the MINOR version goes up, so please could you change the references to 2.5.2 to 2.5.0?  I think then this will just be waiting on sorting out the CreateTime stuff and then it should be good to go.  5:)

RuBublik committed:
fixed build number - resets when MINOR version goes up

RuBublik commented:
Hi there,
@ikelos - thanks for the heads up regarding the versions and build numbers. I have made the necessary changes (:

Regarding the timestamp stuff - as mentioned in my response to @iMHLv2, I investigated the matter but couldn't reproduce the issue. Looking forward for an update from @iMHLv2 regarding the memory sample he used, so I can continue investigating and addressing the issue.

Thank you for your feedback! (:

RuBublik committed:
fixed merge conflicts with 'volatilityfoundation:develop' branch - bumped VERSION_PATCH

RuBublik committed:
added account for XP timestamps - bit shifted

RuBublik committed:
added another sanity check to ETHREAD.is_valid

RuBublik committed:
bumped version constants to mark change of interface (add of support for ETHREAD)

RuBublik committed:
Merge branch 'develop' into poolscanner-thread-support

RuBublik committed:
fix indentation (typo)

RuBublik committed:
Merge branch 'poolscanner-thread-support' of https://github.com/RuBublik/volatility3 into poolscanner-thread-support

ikelos commented:
@iMHLv2 any chance you could give this another look please?

ikelos committed:
Merge branch 'develop' into poolscanner-thread-support

ikelos commented:
@iMHLv2 gave this a look oer and said it's good to go now so in it goes...  5:)
